### ISSUE TITLE
[MJAVADOC-555] Javadoc:aggregate, Javadoc:jar fail if one of the modules does not contain module-info.java

### ISSUE THREAD

**Issue by jira-importer** [apache/maven-javadoc-plugin#849]
**[Gili](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=cowwoc)** opened **[MJAVADOC-555](https://issues.apache.org/jira/browse/MJAVADOC-555?redirect=false)** and commented

1. Unzip testcase
2. Run `mvn clean install javadoc:aggregate`
3. Build fails with: `Exit code: 2 - javadoc: error - No source files for package module2`

Per MPLUGIN-341, Maven plugins cannot contain `module.info.java`. One of my projects builds under Java 11 and is fully modularized except for one module which is a Maven plugin. Due to the aforementioned issue, I cannot use `javadoc:aggregate.`

---

**Affects:** 3.1.0

**Attachments:**
- [javadoc-cannot-link-to-automatic-modules.zip](https://issues.apache.org/jira/secure/attachment/12953787/javadoc-cannot-link-to-automatic-modules.zip) (_2.06 kB_)
- [testcase.zip](https://issues.apache.org/jira/secure/attachment/12952932/testcase.zip) (_3.83 kB_)
- [toolchains.xml](https://issues.apache.org/jira/secure/attachment/12953796/toolchains.xml) (_310 bytes_)

**Issue Links:**
- [MPLUGIN-341](https://issues.apache.org/jira/browse/MPLUGIN-341) Make the plugin tool modular or at least module-ready
 (_**"is blocked by"**_)

**Remote Links:**
- [JDK-8212233
](https://bugs.openjdk.java.net/browse/JDK-8212233)

0 votes, 6 watchers

**Comment by jira-importer** [comment]
**[Robert Scholte](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=rfscholte)** commented

No clue how to solve this yet. The javadoc tool doesn't allow using both source-path and module-source-path.

**Comment by jira-importer** [comment]
**[Gili](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=cowwoc)** commented

[Robert Scholte](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=rfscholte) I suggest the following:
1. Commit a code change that checks for this condition and provides a descriptive error message (possibly pointing to this bug report) if it occurs. This will help developers who might otherwise spend hours trying to debug what is going on.
2. Keep this bug report open until we have a more satisfying conclusion.
3. Publish a release of maven-java-plugin so we begin publishing Java 11 artifacts to Maven Central :)
4. Start a discussion thread on jigsaw-dev seeing how/if we can address this issue.

**Comment by jira-importer** [comment]
**[Robert Scholte](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=rfscholte)** commented

Regarding #4 the solution is probably providing an automatic module name in the manifest file. Must still be implemented though, but I'm not going to spend time on #1 unless somebody provides a patch.

#3 should happen soonish

**Comment by jira-importer** [comment]
**[Gili](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=cowwoc)** commented

[Robert Scholte](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=rfscholte) I think I can work around this issue but there are two other I really need you to merge into the upcoming release: MJAVADOC-554 and MJAVADOC-556. That last one is especially important as I haven't found a workaround for it.

I've just submitted pull requests for both. Assuming these are accepted, when are you planning to cut a release? I'd like to publish a new version of my library to Maven Central and I cannot do so a maven-javadoc-plugin release containing these fixes. Thank you.

**Comment by jira-importer** [comment]
**[Gili](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=cowwoc)** commented

Regarding #4, how would that work exactly? If you look at the testcase, you'll notice that we're running the Javadoc tool against unpacked files. I don't see how you would add a manifest file in this context.

I tried manually adding MANIFEST.MF containing **Automatic-Module-Name: module2** but it did not make a difference.

Either MANIFEST.MF isn't being picked up by the Javadoc tool, or **Automatic-Module-Name** is not sufficient in this case (i.e. we need **module-info.java**).

**Comment by jira-importer** [comment]
**[Gili](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=cowwoc)** commented

`javadoc:jar` is similarly impacted. Reading https://bugs.openjdk.java.net/browse/JDK-8212233 JDK12 and up will not support mixing named and unnamed modules.

Maven plugins cannot use Java modules due to MPLUGIN-341. If I try invoke `javadoc::jar` on a Maven plugin (which is an unnamed module) that depends upon a named module, it fails with the aforementioned error.

Hopefully the fix for `javadoc:aggregate` and `javadoc:jar` will be the same.

**Comment by jira-importer** [comment]
**[Robert Scholte](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=rfscholte)** commented

It is already fixed, see https://builds.apache.org/job/maven-box/job/maven-javadoc-plugin/job/MJAVADOC-555/
I'll wait for MJAVADOC-527 to be merged, then I'll merge this issue.

**Comment by jira-importer** [comment]
**[Gili](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=cowwoc)** commented

[Robert Scholte](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=rfscholte) I just noticed our comments crossed paths. I checked out your MJAVADOC-555 branch and ran it against javadoc-cannot-link-to-automatic-modules.zip but the test still fails. So it seems we've got another loose end to tie up.

**Comment by jira-importer** [comment]
**[Gili](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=cowwoc)** commented

[Robert Scholte](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=rfscholte) I see what you are doing:
* Using --patch-module to compile module2 as if it were a module
* Adding module2 to the list of sources

In theory, this should work for javadoc-cannot-link-to-automatic-modules.zip as well but in practice it does not. Did you try running that testcase against your latest code? It doesn't seem to be doing the above patching. I'm not sure why.

**Comment by jira-importer** [comment]
**[Robert Scholte](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=rfscholte)** commented

Looking closer to the exception: https://google.github.io/guava/releases/27.0.1-jre/api/docs/ is indeed plain old javadoc, it doesn't contain modules as we want to generate here.
I think this is a weird restriction, is there a reason why the external links should be module-based too?

**Comment by jira-importer** [comment]
**[Gili](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=cowwoc)** commented

The JDK issue is saying this behavior is by design so nothing was changed. 

I'm not sure why the restriction is there for external links having to be modules. Are you able to ask Doug from Oracle?

**Comment by jira-importer** [comment]
**[Gili](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=cowwoc)** commented

I just noticed that in the 3rd comment Jonathan Gibson writes "this looks like a Maven issue". Maybe that's why they closed it as Not an Issue. Let me know what you find out.

**Comment by jira-importer** [comment]
**[Robert Scholte](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=rfscholte)** commented

FYI, the JDK issue is reopened and it looks like they found a fix. To be continued...

**Comment by jira-importer** [comment]
**[Robert Scholte](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=rfscholte)** commented

Fixed in [c3e96e622d4ce0ea6e3ba7fcff9cbd2c555ae716](https://gitbox.apache.org/repos/asf?p=maven-javadoc-plugin.git;a=commit;h=c3e96e622d4ce0ea6e3ba7fcff9cbd2c555ae716)
JDK12 has been fixed too, see http://hg.openjdk.java.net/jdk/jdk12/rev/8ce4083fc831

**Comment by jira-importer** [comment]
**[Gili](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=cowwoc)** commented

JDK-8212233 is marked as fixed in JDK 12 b28 but the integration test still fails. Further, the [changelog](https://hg.openjdk.java.net/jdk/jdk/log?rev=reverse%28%22jdk-12%2B27%22%3A%3A%22jdk-12%2B28%22-%22jdk-12%2B27%22%29) doesn't mention JDK-8212233. This leads me to believe that b28 does not contain a fix.

Can you please double-check with Jonathan? I don't have permissions to post on https://bugs.openjdk.java.net/.

**Comment by jira-importer** [comment]
**[Robert Scholte](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=rfscholte)** commented

You need to look in the right branch, it is available at https://hg.openjdk.java.net/jdk/jdk12/

**Comment by jira-importer** [comment]
**[Gili](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=cowwoc)** commented

Okay but the integration test is failing for me under jdk 12 b28. Shouldn't it pass?

**Comment by jira-importer** [comment]
**[Gili](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=cowwoc)** commented

Lets say I skip the integration test for a moment, when I try invoking the "aggregate" goal with:

```java
<javaApiLinks>
 <property>
  <name>api_12</name>
  <value>https://download.java.net/java/early_access/jdk12/docs/api/</value>
 </property>
</javaApiLinks>
```

I get a `NullPointerException` on `AbstractJavadocMojo.getDefaultJavadocApiLink()` because it tries reading `java-api-element-list-12` which does not exist. It looks like you have such files for JDK 1.3 through 11 but nothing for 12. I tried using `<source>11</source>` instead. This did not throw an exception but JDK classes were not linked in the output.

How do I get JDK 12 to build Javadoc that links to the online Java 11 documentation?

**Comment by jira-importer** [comment]
**[Robert Scholte](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=rfscholte)** commented

That's a different issue, related to MJAVADOC-566 and MJAVADOC-562

**Comment by jira-importer** [comment]
**[Gili](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=cowwoc)** commented

Please help me understand what should be possible to run as of today (currently MJAVADOC-562 is marked as fixed and MJAVADOC-566 is not).

Should I be able to run `javadoc:aggregate` under JDK 12 linking to the JDK 12 online Javadoc? Alternatively, should I be able to run `javadoc:aggregate` under JDK 12 linking to the JDK 11 online Javadoc? I don't mind either approach. As it stands, I can't figure out how to mix named and unnamed modules (with an `Automatic-Module-Name` entry).

**UPDATE**: This issue was resolved by a new JDK12 branch that was recently merged into master.
